opnode/l1/downloader.go

```diff
@@ -0,0 +1,271 @@
+package l1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/eth"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+const (
+	fetchBlockTimeout   = 10 * time.Second
+	fetchReceiptTimeout = 10 * time.Second
+	maxReceiptRetry     = 5
+
+	// 500 at 100 KB each would be 50 MB of memory for the L1 block inputs cache
+	cacheSize = 500
+
+	// Amount of receipt tasks to buffer before applying back-pressure (blocking new block-requests)
+	receiptQueueSize = 100
+)
+
+var (
+	DownloadEvictedErr = errors.New("evicted")
+	DownloadClosedErr  = errors.New("closed")
+)
+
+type downloadTask struct {
+	// Incremented when Finish is called, to check whether the task already completed (with or without error).
+	// May increment >1 when other sub-tasks fail.
+	//
+	// First field, aligned atomic changes.
+	finished uint32
+
+	// Count already downloaded receipts to track status
+	//
+	// Aligned after above field, atomic changes.
+	downloadedReceipts uint32
+
+	block *types.Block
+	// receipts slice is allocated in advance, one slot for each transaction, nil until downloaded
+	receipts []*types.Receipt
+
+	// stop fetching if this context is dead
+	ctx context.Context
+	// to free up resources (e.g. ongoing receipt tasks) if a downloadTask finishes prematurely
+	cancel context.CancelFunc
+
+	// feed to subscribe the requests to (de-duplicate work)
+	feed event.Feed
+}
+
+// wrappedErr wraps an error, since event.Feed cannot handle nil errors otherwise (reflection on nil)
+type wrappedErr struct {
+	error
+}
+
+func (bl *downloadTask) Finish(err wrappedErr) {
+	if atomic.AddUint32(&bl.finished, 1) == 1 {
+		bl.feed.Send(err)
+	}
+}
+
+type receiptTask struct {
+	blockHash common.Hash
+	txHash    common.Hash
+	txIndex   uint64
+	// Count the attempts we made to fetch this receipt. Block as a whole fails if we tried too many times.
+	retry uint64
+	// Avoid concurrent Downloader cache access and pruning edge cases with receipts.
+	// Keep a pointer to the parent task to insert the receipt into.
+	dest *downloadTask
+}
+
+type Downloader interface {
+	Fetch(ctx context.Context, id eth.BlockID) (*types.Block, []*types.Receipt, error)
+	AddReceiptWorkers(n int) int
+	Close()
+}
+
+type DownloadSource interface {
+	eth.BlockByHashSource
+	eth.ReceiptSource
+}
+
+type downloader struct {
+	// cache of ongoing/completed block tasks: block hash -> block
+	cacheEvict *lru.Cache
+	cacheLock  sync.Mutex
```

Comment:
The LRU cache itself is thread safe, but this lock seems to be protecting against duplicate requests.

Reply:
Indeed, the lock makes sure we are not concurrently adding a new duplicate job.

Consider this race:
```
routine A and B get from the cache (in parallel, but safe)
routine A and B find nothing
routine A and B create and add a new job
```
The LRU interface does provide a `ContainsOrAdd`, but that doesn't update the frequency like `Get`, and forces us to allocate before knowing it's a duplicate or not.


I'll try add a comment about the atomic `Get / Add` to avoid duplicate jobs.

Reply:
Good point, is that actually necessary?